src/bt-core-io.c

```diff
@@ -507,31 +506,65 @@ alarm_event(int fd, uint32_t events, void* data)
   return res;
 }
 
-static bool
-set_wake_alarm_cb(uint64_t delay_millis, bool should_wake, alarm_cb cb,
-                  void* data)
+struct set_wake_alarm_param {
+  uint64_t delay_millis;
+  bool should_wake;
+  alarm_cb cb;
+  void* data;
+};
+
+static enum ioresult
+set_wake_alarm_task_cb(void* data)
 {
+  struct set_wake_alarm_param* param = data;
   struct alarm_state* alarm_state;
   int clockid;
+  assert(param);
 
-  errno = 0;
   alarm_state = malloc(sizeof(*alarm_state));
-  if (errno) {
+  if (!alarm_state) {
     ALOGE_ERRNO("malloc");
-    return false;
+    goto err_malloc_alarm_state;
   }
+  alarm_state->cb = param->cb;
+  alarm_state->data = param->data;
 
-  alarm_state->cb = cb;
-  alarm_state->data = data;
+  clockid = param->should_wake ? CLOCK_BOOTTIME_ALARM : CLOCK_BOOTTIME;
+  if (add_relative_timer_to_epoll_loop(clockid, param->delay_millis, 0,
+                                       alarm_event, alarm_state) < 0)
+    goto err_add_relative_timer_to_epoll_loop;
 
-  clockid = should_wake ? CLOCK_BOOTTIME_ALARM : CLOCK_BOOTTIME;
-  if (add_relative_timer_to_epoll_loop(clockid, delay_millis, 0, alarm_event,
-                                       alarm_state) < 0)
-    goto err_add_timer_to_epoll_loop;
+  free(param);
+  return IO_OK;
+err_add_relative_timer_to_epoll_loop:
+  free(alarm_state);
+err_malloc_alarm_state:
+  free(param);
+  return IO_OK;
+}
+
+static bool
+set_wake_alarm_cb(uint64_t delay_millis, bool should_wake, alarm_cb cb,
+                  void* data)
+{
+  struct set_wake_alarm_param* param;
+
+  param = malloc(sizeof(*param));
+  if (!param) {
```

Comment:
Good :) Unfortunately we found a platform where errno gets set by successful mallocs. So testing for !NULL is the better option. I'm a bit sad about this. ;)